---
title: <a id="Mutation Report"></a> <center> <h1>Mutation Report</h1> </center>
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
args <- commandArgs(trailingOnly = TRUE)
sample.name <- args[7]
```

---
pagetitle: Report of `r sample.name`
---

```{r, include=FALSE}
source("imports.R", local = knitr::knit_global())
```

 Patient information
=====================================================

```{r Patient_Information, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
##Patients information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
if (args[10] == "tumnorm")
{
  sample.type <- "Case and Control Biopsy"
} else {
  sample.type <- "Case Biopsy"
}
pat <- data.frame(Name = args[1], Surname = args[2], ID = args[3], Gender = args[4], Age = args[5],
                  Sample = sample.type, Tumor = args[6])
kable(pat) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```


## Detected variant therapeutic benefit  {.tabset}

### Variant Mutation
```{r Variant_Mutation, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], ".txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  x_trial <- read.csv(paste0(args[8], "/txt/trial/", args[7], ".txt"), sep = "\t")
  list.all.trials <- unique(x_trial$Clinical_trial)

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_definitive.txt"), sep = "\t", colClasses = c("character"))
  dis <- read.csv(paste0(args[9], "/Disease.txt"), sep = "\t")
  x <- merge(dis, x, by = "Disease")
  x$Disease <- NULL
  colnames(x)[1] <- "Disease"
  x[is.na(x)] <- " "
  x <- x[x$Evidence_direction == "Supports" & x$Disease == args[6], , drop = F]
  x$Drug <- as.character(x$Drug, levels = (x$Drug))
  x$Drug <- gsub(", ", ",", x$Drug, fixed = T)
  x$Drug_interaction_type <- gsub(" ", "", x$Drug_interaction_type)
  x$Evidence_type <- factor(x$Evidence_type, levels = c("Diagnostic", "Prognostic", "Predisposing", "Predictive"))
  x$Evidence_level <- factor(x$Evidence_level, levels = c("Validated association", "FDA guidelines", "NCCN guidelines", "Clinical evidence", "Late trials", "Early trials", "Case study", "Case report", "Preclinical evidence", "Pre-clinical", "Inferential association"))

  x <- inner_join(x, x_url)
  x <- x[order(x$Gene, x$Evidence_level, x$Evidence_type, x$Variant, x$Drug, x$Drug_interaction_type,
               x$Clinical_significance, x$Reference),]
  if (nrow(x) > 0)
    x$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', x$Reference, '">', x$Reference, '</a>')

  hgx <- split(x, x$Gene)
  empty <- T
  for (n in hgx)
  {
    if (nrow(n) > 0)
    {
      gene <- as.character(n$Gene)[1]
      cat("  \n#### Gene", gene, " \n")
      n <- n[, c("Variant", "Drug", "Drug_interaction_type", "Evidence_type", "Clinical_significance", "Evidence_level", "Evidence_statement", "Type", "Reference")]
      tmp <- group_by(n, Evidence_level, Evidence_type, Variant, Drug, Drug_interaction_type, Clinical_significance, Type)
      tmp2 <- summarise(tmp, Evidence_statement = paste(Evidence_statement, collapse = " "), Reference = paste(sort(Reference), collapse = ", "))
      ya <- tmp2[, c("Variant", "Drug", "Drug_interaction_type", "Evidence_type", "Clinical_significance", "Evidence_level", "Type", "Reference"), drop = F]
      yb <- tmp2[, c("Evidence_statement", "Reference"), drop = F]
      ya$Evidence <- 1:nrow(ya)
      ya$Details <- paste0('[<a href="Javascript:;" class="my_ref_link" data-id="#det-', gene, '-', ya$Evidence, '">+</a>]')
      ya$Details[!complete.cases(yb) |
                   yb$Evidence_statement == " " |
                   yb$Evidence_statement == ""] <- ""
      ya$Evidence <- paste0('<a id="evi-', gene, '-', ya$Evidence, '" name="evi-', gene, '-', ya$Evidence, '"></a>', ya$Evidence)
      list.trials <- paste0("https://clinicaltrials.gov/ct2/results?cond=", ya$Variant, "&term=",
                            ya$Drug, "&cntry=&state=&city=&dist=")
      ya$Trials <- paste0("[", "<a href=\"", list.trials, "\">+</a>", "]")
      ya$Trials[!list.trials %in% list.all.trials] <- ""
      ya$Drug <- gsub(",", ", ", ya$Drug, fixed = T)
      ya <- ya[, c("Evidence", "Variant", "Drug", "Drug_interaction_type", "Evidence_type", "Clinical_significance", "Evidence_level", "Type", "Details", "Trials", "Reference")]
      names(ya) <- c("#", "Variant", "Drug", "Drug Interaction", "Evidence Type", "Clinical Significance",
                     "Evidence Level", "Type", "Details", "Trials", "References")
      if (args[10] == "tumnorm")
        ya <- ya[, -which(names(ya) == "Type")]
      print(ya %>%
              kable("html", escape = FALSE) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```

### Evidence Details
```{r Lead_disease_evidence_details, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], ".txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_definitive.txt"), sep = "\t", colClasses = c("character"))
  dis <- read.csv(paste0(args[9], "/Disease.txt"), sep = "\t")
  x <- merge(dis, x, by = "Disease")
  x$Disease <- NULL
  colnames(x)[1] <- "Disease"
  x[is.na(x)] <- " "
  x <- x[x$Evidence_direction == "Supports" & x$Disease == args[6], , drop = F]
  x$Drug <- as.character(x$Drug, levels = (x$Drug))
  x$Drug_interaction_type <- gsub(" ", "", x$Drug_interaction_type)
  x$Evidence_type <- factor(x$Evidence_type, levels = c("Diagnostic", "Prognostic", "Predisposing", "Predictive"))
  x$Evidence_level <- factor(x$Evidence_level, levels = c("Validated association", "FDA guidelines", "NCCN guidelines", "Clinical evidence", "Late trials", "Early trials", "Case study", "Case report", "Preclinical evidence", "Pre-clinical", "Inferential association"))

  x <- inner_join(x, x_url)
  x <- x[order(x$Gene, x$Evidence_level, x$Evidence_type, x$Variant, x$Drug, x$Drug_interaction_type,
               x$Clinical_significance, x$Reference),]
  if (nrow(x) > 0)
    x$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', x$Reference, '">', x$Reference, '</a>')

  hgx <- split(x, x$Gene)
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      gene <- as.character(n$Gene)[1]
      cat("  \n#### Gene", gene, " \n")
      n2 <- n[, c("Variant", "Drug", "Drug_interaction_type", "Evidence_type", "Clinical_significance", "Evidence_level", "Evidence_statement", "Reference")]
      tmp <- group_by(n2, Evidence_level, Evidence_type, Variant, Drug, Drug_interaction_type, Clinical_significance)
      tmp2 <- summarise(tmp, Evidence_statement = paste(Evidence_statement, collapse = " "),
                        Reference = paste(sort(Reference), collapse = ", "))
      yb <- tmp2[, c("Evidence_statement", "Reference"), drop = F]
      yb$Evidence <- 1:nrow(yb)
      yb$Less <- paste0('[<a href="Javascript:;" class="my_ref_link" data-id="#evi-', gene, '-', yb$Evidence, '">-</a>]')
      yb$Evidence <- paste0('<a id="det-', gene, '-', yb$Evidence, '" name="det-', gene, '-', yb$Evidence, '"></a>', yb$Evidence)
      yb <- yb[complete.cases(yb) &
                 yb$Evidence_statement != " " &
                 yb$Evidence_statement != "", , drop = F]
      if (nrow(yb) > 0)
      {
        yb <- yb[, c("Evidence", "Evidence_statement", "Reference", "Less"), drop = F]
        names(yb) <- c("#", "Evidence Statement", "References", "")
        print(kable(yb, "html", escape = FALSE) %>%
                kable_styling(bootstrap_options = c("striped", "hover", "responsive", align = "justify")))
      }
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```

### Variant Details
```{r Lead_disease_variant_details, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_definitive.txt"), sep = "\t", colClasses = c("character"))
  dis <- read.csv(paste0(args[9], "/Disease.txt"), sep = "\t")
  x <- merge(dis, x, by = "Disease")
  x$Disease <- NULL
  colnames(x)[1] <- "Disease"
  x[is.na(x)] <- " "
  x <- x[x$Evidence_direction == "Supports" & x$Disease == args[6], , drop = F]
  x$Var_base <- as.character(x$Var_base)
  x$Ref_base <- as.character(x$Ref_base)
  x[is.na(x$Var_base), "Var_base"] <- "T"
  x[is.na(x$Ref_base), "Ref_base"] <- "T"
  x <- x[order(x$Gene, x$Variant),]

  hgx <- split(x, x$Gene)
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      gene <- as.character(n$Gene)[1]
      cat("  \n#### Gene", gene, " \n")
      ya <- n[, c("Variant", "Chromosome", "Ref_base", "Var_base", "Start", "Stop")]
      ya <- unique(ya)
      #ya$Variant <- paste0('<a id="var-',gene,'-',ya$Variant,'" name="var-',gene,'-',ya$Variant,'"></a>',ya$Variant)
      row.names(ya) <- NULL
      print(kable(ya, "html", escape = FALSE) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))

      yc <- n[, c("Variant", "Variant_summary"), drop = F]
      yc <- yc[complete.cases(yc),]
      yc <- unique(yc)
      yc <- yc[!(yc$Variant_summary == " "),]
      yc <- yc[!(yc$Variant_summary == ""),]
      if (nrow(yc) > 0)
      {
        tmp <- group_by(yc, Variant)
        tmp2 <- summarise(tmp, Variant_summary = paste(Variant_summary, collapse = " "))
        print(kable(tmp2) %>%
                kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      }

      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```


## PharmGKB variant {.tabset}

### Variant Mutation
```{r pharm_variant, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_pharm.txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_pharm.txt"), sep = "\t", colClasses = c("character"))
  x[is.na(x)] <- " "
  x$Drug <- as.character(x$Drug, levels = (x$Drug))
  x$Drug <- gsub(", ", ",", x$Drug, fixed = T)
  x$Gene <- as.character(x$Gene, levels = (x$Gene))
  x$Clinical_significance <- gsub(pattern = " ", replace = "", x$Clinical_significance)

  x <- inner_join(x, x_url)
  x <- x[order(x$Gene, x$Variant, x$Drug, x$Clinical_significance, x$Reference),]
  if (nrow(x) > 0)
  {
    x$Reference <- paste0(x$Reference, "p")
    x$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', x$Reference, '">', x$Reference, '</a>')
  }
  hgx <- split(x, paste(x$Gene))
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      gene <- as.character(n$Gene)[1]
      cat("  \n#### Gene", gene, " \n")
      n <- n[, c("Variant", "Drug", "Clinical_significance", "Evidence_statement", "Type", "Reference")]
      tmp <- group_by(n, Variant, Drug, Clinical_significance, Type)
      tmp2 <- summarise(tmp, Evidence_statement = paste(Evidence_statement, collapse = " "), Reference = paste(Reference, collapse = ", "))
      ya <- tmp2[, c("Variant", "Drug", "Clinical_significance", "Type", "Reference"), drop = F]
      yb <- tmp2[, c("Evidence_statement", "Reference"), drop = F]
      ya$Evidence <- 1:nrow(ya)
      ya$Details <- paste0('[<a href="Javascript:;" class="my_ref_link" data-id="#det-', gene, '-', ya$Evidence, 'p">+</a>]')
      ya$Details[!complete.cases(yb) |
                   yb$Evidence_statement == " " |
                   yb$Evidence_statement == ""] <- ""
      ya$Evidence <- paste0('<a id="evi-', gene, '-', ya$Evidence, 'p" name="evi-', gene, '-', ya$Evidence, 'p"></a>', ya$Evidence)
      ya$Drug <- gsub(",", ", ", ya$Drug, fixed = T)
      ya <- ya[, c("Evidence", "Variant", "Drug", "Clinical_significance", "Type", "Details", "Reference")]
      names(ya) <- c("#", "Variant", "Drug", "Clinical Significance", "Type", "Details", "References")
      if (args[10] == "tumnorm")
        ya <- ya[, -which(names(ya) == "Type")]
      print(kable(ya, "html", escape = FALSE) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```

### Evidence Details
```{r pharm_evidence_details, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_pharm.txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_pharm.txt"), sep = "\t", colClasses = c("character"))
  x[is.na(x)] <- " "
  x$Drug <- as.character(x$Drug, levels = (x$Drug))
  x$Gene <- as.character(x$Gene, levels = (x$Gene))
  x$Clinical_significance <- gsub(pattern = " ", replace = "", x$Clinical_significance)

  x <- inner_join(x, x_url)
  x <- x[order(x$Gene, x$Variant, x$Drug, x$Clinical_significance, x$Reference),]
  if (nrow(x) > 0)
  {
    x$Reference <- paste0(x$Reference, "p")
    x$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', x$Reference, '">', x$Reference, '</a>')
  }
  hgx <- split(x, paste(x$Gene))
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      gene <- as.character(n$Gene)[1]
      cat("  \n#### Gene", gene, " \n")
      n2 <- n[, c("Variant", "Drug", "Clinical_significance", "Evidence_statement", "Reference")]
      tmp <- group_by(n2, Variant, Drug, Clinical_significance)
      tmp2 <- summarise(tmp, Evidence_statement = paste(Evidence_statement, collapse = " "),
                        Reference = paste(sort(Reference), collapse = ", "))
      yb <- tmp2[, c("Evidence_statement", "Reference"), drop = F]
      yb$Evidence <- 1:nrow(yb)
      yb$Less <- paste0('[<a href="Javascript:;" class="my_ref_link" data-id="#evi-', gene, '-', yb$Evidence, 'p">-</a>]')
      yb$Evidence <- paste0('<a id="det-', gene, '-', yb$Evidence, 'p" name="det-', gene, '-', yb$Evidence, 'p"></a>', yb$Evidence)
      yb <- yb[complete.cases(yb) &
                 yb$Evidence_statement != " " &
                 yb$Evidence_statement != "", , drop = F]
      if (nrow(yb) > 0)
      {
        yb <- yb[, c("Evidence", "Evidence_statement", "Reference", "Less"), drop = F]
        names(yb) <- c("#", "Evidence Statement", "References", "")
        print(kable(yb, "html", escape = FALSE) %>%
                kable_styling(bootstrap_options = c("striped", "hover", "responsive", align = "justify")))
      }
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```

### Variant Details
```{r pharm_variant_details, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_pharm.txt"), sep = "\t", colClasses = c("character"))
  x[is.na(x)] <- " "
  x$Gene <- as.character(x$Gene, levels = (x$Gene))
  x$Var_base <- as.character(x$Var_base)
  x$Ref_base <- as.character(x$Ref_base)
  x[is.na(x$Var_base), "Var_base"] <- "T"
  x[is.na(x$Ref_base), "Ref_base"] <- "T"

  x <- inner_join(x, x_url)
  x <- x[order(x$Gene, x$Variant),]

  hgx <- split(x, paste(x$Gene))
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      cat("  \n#### Gene", as.character(n$Gene)[1], " \n")
      ya <- n[, c("Variant", "Chromosome", "Ref_base", "Var_base", "Start", "Stop")]
      ya <- unique(ya)
      row.names(ya) <- NULL
      print(kable(ya[1,]) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      yc <- n[, c("Variant", "Variant_summary"), drop = F]
      yc <- yc[complete.cases(yc),]
      yc <- unique(yc)
      yc <- yc[!(yc$Variant_summary == " "),]
      yc <- yc[!(yc$Variant_summary == ""),]
      if (nrow(yc) > 0)
      {
        tmp <- group_by(yc, Variant)
        tmp2 <- summarise(tmp, Variant_summary = paste(Variant_summary, collapse = " "))
        print(kable(tmp2) %>%
                kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      }
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```


## Mutations' annotations
```{r mut_annotations, echo=FALSE, message=FALSE, warning=FALSE}
args <- commandArgs(trailingOnly = TRUE)
try({
  xref <- read.csv(paste0(args[8], "/txt/", args[7], "_refgene.txt"), sep = "\t", colClasses = c("character"))
  xclin <- read.csv(paste0(args[8], "/txt/", args[7], "_clinvar.txt"), sep = "\t", colClasses = c("character"))
  xcr <- merge(xref, xclin, all.x = TRUE)
  xcr <- format.data.frame(xcr, digits = NULL, na.encode = FALSE)
  xcr[is.na(xcr)] <- " "
  xcr <- xcr[, c("Gene", "Chromosome", "Stop", "Ref_base", "Var_base", "Change_type", "Clinical_significance", "Type")]
  names(xcr) <- c("Gene", "Chromosome", "Position", "Ref_base", "Var_base",
                  "Change Type", "Clinical Significance", "Type")
  if (args[10] != "tumnorm")
    xcr$Type <- factor(xcr$Type, levels = c("Somatic", "Germline"))
  xcr <- xcr[order(xcr$Type, xcr$Gene, as.integer(xcr$Position)),]
  rownames(xcr) <- NULL
  if (nrow(xcr) > 0)
  {
    if (args[10] == "tumnorm")
      xcr <- xcr[, -which(names(xcr) == "Type")]
    kable(xcr) %>%
      kable_styling(bootstrap_options = c("striped", "hover", "responsive")) %>%
      column_spec(1, bold = T, border_right = T)
  } else
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```


## Off labels Drug  {.tabset}

### Variant Mutation
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_off.txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  x_trial <- read.csv(paste0(args[8], "/txt/trial/", args[7], "_off.txt"), sep = "\t")
  list.all.trials <- unique(x_trial$Clinical_trial)

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_definitive.txt"), sep = "\t", colClasses = c("character"))
  dis <- read.csv(paste0(args[9], "/Disease.txt"), sep = "\t")
  x <- merge(dis, x, by = "Disease")
  x$Disease <- NULL
  colnames(x)[1] <- "Disease"
  x <- x[x$Evidence_direction == "Supports" & x$Disease != args[6], , drop = F]
  x[is.na(x)] <- " "
  x$Drug <- as.character(x$Drug, levels = (x$Drug))
  x$Drug <- gsub(", ", ",", x$Drug, fixed = T)
  x$Evidence_type <- factor(x$Evidence_type, levels = c("Diagnostic", "Prognostic", "Predisposing", "Predictive"))
  x$Evidence_level <- factor(x$Evidence_level, levels = c("Validated association", "FDA guidelines", "NCCN guidelines", "Clinical evidence", "Late trials", "Early trials", "Case study", "Case report", "Preclinical evidence", "Pre-clinical", "Inferential association"))
  x$Disease <- as.character(x$Disease, levels = (x$Disease))

  x <- inner_join(x, x_url)
  x <- x[order(x$Disease, x$Evidence_level, x$Evidence_type, x$Gene, x$Variant, x$Drug,
               x$Clinical_significance, x$Reference),]
  if (nrow(x) > 0)
  {
    x$Reference <- paste0(x$Reference, "a")
    x$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', x$Reference, '">', x$Reference, '</a>')
    x <- x[, c("Disease", "Gene", "Variant", "Drug", "Evidence_type", "Clinical_significance", "Evidence_level", "Evidence_statement", "Type", "Reference")]
    tmp <- group_by(x, Disease, Evidence_level, Evidence_type, Gene, Variant, Drug, Clinical_significance, Type)
    tmp2 <- summarise(tmp, Evidence_statement = paste(Evidence_statement, collapse = " "), Reference = paste(Reference, collapse = ", "))

    ya <- tmp2[, c("Disease", "Gene", "Variant", "Drug", "Evidence_type", "Clinical_significance", "Evidence_level", "Type", "Reference"), drop = F]
    yb <- tmp2[, c("Evidence_statement", "Reference"), drop = F]
    tmp <- unname(unlist(lapply(split(ya, ya$Disease), nrow)))
    ya$Evidence <- unlist(sapply(tmp, seq, from = 1))
    dis.without.spaces <- gsub(" ", "", ya$Disease)
    ya$Details <- paste0('[<a href="Javascript:;" class="my_ref_link" data-id="#det-', dis.without.spaces, '-', ya$Evidence, 'a">+</a>]')
    ya$Details[!complete.cases(yb) |
                 yb$Evidence_statement == " " |
                 yb$Evidence_statement == ""] <- ""
    ya$Evidence <- paste0('<a id="evi-', dis.without.spaces, '-', ya$Evidence, 'a" name="evi-', dis.without.spaces, '-', ya$Evidence, 'a"></a>', ya$Evidence)
    list.trials <- paste0("https://clinicaltrials.gov/ct2/results?cond=", ya$Variant, "&term=",
                          ya$Drug, "&cntry=&state=&city=&dist=")
    ya$Trials <- paste0("[", "<a href=\"", list.trials, "\">+</a>", "]")
    ya$Trials[!list.trials %in% list.all.trials] <- ""
    ya$Drug <- gsub(",", ", ", ya$Drug, fixed = T)
    ya <- ya[, c("Evidence", "Disease", "Gene", "Variant", "Drug", "Evidence_type", "Clinical_significance", "Evidence_level", "Type", "Details", "Trials", "Reference")]
    names(ya) <- c("#", "Disease", "Gene", "Variant", "Drug", "Evidence Type", "Clinical Significance",
                   "Evidence Level", "Type", "Details", "Trials", "References")
    if (args[10] == "tumnorm")
      ya <- ya[, -which(names(ya) == "Type")]
    df_total <- ya[, -which(names(ya) == "Disease")]
    if (nrow(df_total) > 0)
    {
      t <- kable(df_total, "html", escape = F) %>%
        kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
      ya$Disease <- as.factor(ya$Disease)
      lvl <- levels(ya$Disease)
      for (l in lvl)
      {
        a <- which(ya$Disease == l)
        # if (length(a)!=0)
        t <- t %>% pack_rows(l, min(a), max(a), indent = FALSE)
      }
      print(t)
    } else
      cat("  \n##### No data available.  \n")
  } else
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```

### Evidence Details
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_off.txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_definitive.txt"), sep = "\t", colClasses = c("character"))
  dis <- read.csv(paste0(args[9], "/Disease.txt"), sep = "\t")
  x <- merge(dis, x, by = "Disease")
  x$Disease <- NULL
  colnames(x)[1] <- "Disease"
  x[is.na(x)] <- " "
  x <- x[x$Evidence_direction == "Supports" & x$Disease != args[6], , drop = F]
  x$Drug <- as.character(x$Drug, levels = (x$Drug))
  x$Evidence_type <- factor(x$Evidence_type, levels = c("Diagnostic", "Prognostic", "Predisposing", "Predictive"))
  x$Evidence_level <- factor(x$Evidence_level, levels = c("Validated association", "FDA guidelines", "NCCN guidelines", "Clinical evidence", "Late trials", "Early trials", "Case study", "Case report", "Preclinical evidence", "Pre-clinical", "Inferential association"))
  x$Disease <- as.character(x$Disease, levels = (x$Disease))

  x <- inner_join(x, x_url)
  x <- x[order(x$Disease, x$Evidence_level, x$Evidence_type, x$Gene, x$Variant, x$Drug,
               x$Clinical_significance, x$Reference),]
  if (nrow(x) > 0)
  {
    x$Reference <- paste0(x$Reference, "a")
    x$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', x$Reference, '">', x$Reference, '</a>')
    x2 <- x[, c("Disease", "Gene", "Variant", "Drug", "Evidence_type", "Clinical_significance", "Evidence_level", "Evidence_statement", "Reference")]
    tmp <- group_by(x2, Disease, Evidence_level, Evidence_type, Gene, Variant, Drug, Clinical_significance)
    tmp2 <- summarise(tmp, Evidence_statement = paste(Evidence_statement, collapse = " "),
                      Reference = paste(sort(Reference), collapse = ", "))
    yb <- tmp2[, c("Disease", "Evidence_statement", "Reference"), drop = F]
    tmp <- unname(unlist(lapply(split(yb, yb$Disease), nrow)))
    yb$Evidence <- unlist(sapply(tmp, seq, from = 1))
    dis.without.spaces <- gsub(" ", "", yb$Disease)
    yb$Less <- paste0('[<a href="Javascript:;" class="my_ref_link" data-id="#evi-', dis.without.spaces, '-', yb$Evidence, 'a">-</a>]')
    yb$Evidence <- paste0('<a id="det-', dis.without.spaces, '-', yb$Evidence, 'a" name="det-', dis.without.spaces, '-', yb$Evidence, 'a"></a>', yb$Evidence)
    yb <- yb[complete.cases(yb) &
               yb$Evidence_statement != " " &
               yb$Evidence_statement != "", , drop = F]
    if (nrow(yb) > 0)
    {
      yb2 <- yb[, c("Evidence", "Evidence_statement", "Reference", "Less"), drop = F]
      names(yb2) <- c("#", "Evidence Statement", "References", "")
      t <- kable(yb2, "html", escape = FALSE) %>%
        kable_styling(bootstrap_options = c("striped", "hover", "responsive", align = "justify"))
      yb$Disease <- as.factor(yb$Disease)
      lvl <- levels(yb$Disease)
      for (l in lvl)
      {
        a <- which(yb$Disease == l)
        t <- t %>% pack_rows(l, min(a), max(a), indent = FALSE)
      }
      print(t)
    } else
      cat("  \n##### No data available.  \n")
  } else
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```

### Variant Details
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({

  x <- read.csv(paste0(args[8], "/txt/", args[7], "_definitive.txt"), sep = "\t", colClasses = c("character"))
  dis <- read.csv(paste0(args[9], "/Disease.txt"), sep = "\t")
  x <- merge(dis, x, by = "Disease")
  x$Disease <- NULL
  colnames(x)[1] <- "Disease"
  x[is.na(x)] <- " "
  x <- x[x$Evidence_direction == "Supports" & x$Disease != args[6], , drop = F]
  x$Var_base <- as.character(x$Var_base)
  x$Ref_base <- as.character(x$Ref_base)
  x[is.na(x$Var_base), "Var_base"] <- "T"
  x[is.na(x$Ref_base), "Ref_base"] <- "T"
  x <- x[order(x$Gene, x$Variant),]

  if (nrow(x) > 0)
  {
    ya <- x[, c("Gene", "Variant", "Chromosome", "Ref_base", "Var_base", "Start", "Stop")]
    ya <- unique(ya)
    row.names(ya) <- NULL
    print(kable(ya) %>%
            kable_styling(bootstrap_options = c("striped", "hover", "responsive")))

    yc <- x[, c("Gene", "Variant", "Variant_summary"), drop = F]
    yc <- yc[complete.cases(yc),]
    yc <- unique(yc)
    yc <- yc[!(yc$Variant_summary == " "),]
    yc <- yc[!(yc$Variant_summary == ""),]
    if (nrow(yc) > 0)
    {
      tmp <- group_by(yc, Gene, Variant)
      tmp2 <- summarise(tmp, Variant_summary = paste(Variant_summary, collapse = " "))
      print(kable(tmp2) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
    }
  } else
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```


## Cosmic {.tabset}

### Drug resistant mutations
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_cosmic.txt"), sep = "\t")
  x_url <- x_url[, c("PMID", "Reference")]
  x_url$PMID <- as.character(x_url$PMID)

  cosm <- read.csv(paste0(args[8], "/txt/", args[7], "_cosmic.txt"), sep = "\t", colClasses = c("character"))
  cosm <- cosm[, c("Gene", "Variant", "Drug", "Primary.Tissue", "PMID")]
  cosm <- unique(cosm)

  cosm <- inner_join(cosm, x_url)
  cosm <- cosm[order(cosm$Gene, cosm$Variant, cosm$Drug, cosm$Primary.Tissue, cosm$Reference),]
  if (nrow(cosm) > 0)
  {
    cosm$Reference <- paste0(cosm$Reference, "c")
    cosm$Reference <- paste0('<a href="Javascript:;" class="my_ref_link" data-id="#ref-', cosm$Reference, '">', cosm$Reference, '</a>')
  }
  hgx <- split(cosm, paste(cosm$Gene))
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      cat("  \n#### Gene", as.character(n$Gene)[1], " \n")
      n <- n[, c("Variant", "Drug", "Primary.Tissue", "Reference")]
      tmp <- group_by(n, Variant, Drug, Primary.Tissue)
      tmp2 <- summarise(tmp, Reference = paste(Reference, collapse = ","))
      tmp2$Number <- 1:nrow(tmp2)
      tmp2$Number <- as.character(tmp2$Number)
      tmp2 <- tmp2[, c("Number", "Variant", "Drug", "Primary.Tissue", "Reference")]
      names(tmp2) <- c("#", "Variant", "Drug", "Primary Tissue", "References")
      print(kable(tmp2, "html", escape = F) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)

```

### Variant Details
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
##Mutations information
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x <- read.csv(paste0(args[8], "/txt/", args[7], "_cosmic.txt"), sep = "\t", colClasses = c("character"))
  x[is.na(x$Var_base), "Var_base"] <- "T"
  x[is.na(x$Ref_base), "Ref_base"] <- "T"
  x <- x[order(x$Gene, x$Variant),]
  hgx <- split(x, x$Gene)
  empty <- T
  for (n in hgx)
  {
    if (dim(n)[1] != 0)
    {
      cat("  \n#### Gene", as.character(n$Gene)[1], " \n")
      ya <- n[, c("Variant", "Chromosome", "Ref_base", "Var_base", "Start", "Stop")]
      ya <- unique(ya)
      row.names(ya) <- NULL
      print(kable(ya) %>%
              kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
      empty <- F
    }
  }
  if (empty)
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```


## Drug-Food Interactions
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#Food interactions
args <- commandArgs(trailingOnly = TRUE)
try({
  k <- read.csv(paste0(args[8], "/txt/", args[7], "_drugfood.txt"), sep = "\t")
  k <- k[, c("Drug", "Food_interaction")]
  if (nrow(k) > 0)
  {
    k <- unique(k)
    tmp <- group_by(k, Drug)
    tmp2 <- summarise(tmp, Food_interaction = paste(Food_interaction, collapse = " "))
    names(tmp2) <- c("Drug", "Food Interaction")
    print(kable(tmp2) %>%
            kable_styling(bootstrap_options = c("striped", "hover", "responsive")))
  } else
    cat("  \n##### No data available.  \n")
}, silent = TRUE)
```


## Reference {.tabset}

### Mutation
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], ".txt"), sep = "\t")
  link1 <- x_url[, c("Reference", "PMID", "Citation")]
  if (nrow(link1) > 0)
  {
    link1$Reference <- as.character(link1$Reference)
    link1$Reference <- paste0('<a id="ref-', link1$Reference, '" name="ref-', link1$Reference, '"></a>', link1$Reference)
    t <- link1 %>%
      mutate(PMID = cell_spec(PMID, "html", link = x_url$URL)) %>%
      kable("html", escape = FALSE) %>%
      kable_styling(bootstrap_options = c("hover", "condensed"))
    x_url$Gene <- as.factor(x_url$Gene)
    lvl <- levels(x_url$Gene)
    for (l in lvl)
    {
      a <- which(x_url$Gene == l)
      if (length(a) != 0)
        t <- t %>% pack_rows(l, min(a), max(a), indent = FALSE)
    }
    print(t)
  } else
    cat("  \n##### No references available.  \n")
}, silent = TRUE)
```

### Pharm
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
#pharm
try({
  x_url_p <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_pharm.txt"), sep = "\t")
  link1 <- x_url_p[, c("Reference", "PMID")]
  if (nrow(link1) > 0)
  {
    link1$Reference <- paste0(link1$Reference, "p")
    link1$Reference <- paste0('<a id="ref-', link1$Reference, '" name="ref-', link1$Reference, '"></a>', link1$Reference)
    t <- link1 %>%
      mutate(PMID = cell_spec(PMID, "html", link = x_url_p$URL)) %>%
      kable("html", escape = FALSE) %>%
      kable_styling(bootstrap_options = c("hover", "condensed"))
    x_url_p$Gene <- as.factor(x_url_p$Gene)
    lvl <- levels(x_url_p$Gene)
    for (l in lvl)
    {
      a <- which(x_url_p$Gene == l)
      if (length(a) != 0)
        t <- t %>% pack_rows(l, min(a), max(a), indent = FALSE)
    }
    print(t)
  } else
    cat("  \n##### No references available.  \n")
}, silent = TRUE)

```

### Off label Drug
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
#Off label
try({
  x_url <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_off.txt"), sep = "\t")
  link1 <- x_url[, c("Reference", "PMID", "Citation")]
  if (nrow(link1) > 0)
  {
    link1$Reference <- paste0(link1$Reference, "a")
    link1$Reference <- paste0('<a id="ref-', link1$Reference, '" name="ref-', link1$Reference, '"></a>', link1$Reference)
    t <- link1 %>%
      mutate(PMID = cell_spec(PMID, "html", link = x_url$URL)) %>%
      kable("html", escape = FALSE) %>%
      kable_styling(bootstrap_options = c("hover", "condensed"))
    x_url$Disease <- as.factor(x_url$Disease)
    lvl <- levels(x_url$Disease)
    for (l in lvl)
    {
      a <- which(x_url$Disease == l)
      if (length(a) != 0)
        t <- t %>% pack_rows(l, min(a), max(a), indent = FALSE)
    }
    print(t)
  } else
    cat("  \n##### No references available.  \n")
}, silent = TRUE)
```

### Cosmic
```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
options(knitr.table.format = "html")
args <- commandArgs(trailingOnly = TRUE)
#Cosmic
try({
  cosm <- read.csv(paste0(args[8], "/txt/reference/", args[7], "_cosmic.txt"), sep = "\t")
  link1 <- cosm[, c("Reference", "PMID")]
  if (nrow(link1) > 0)
  {
    link1$Reference <- paste0(link1$Reference, "c")
    link1$Reference <- paste0('<a id="ref-', link1$Reference, '" name="ref-', link1$Reference, '"></a>', link1$Reference)
    t <- link1 %>%
      mutate(PMID = cell_spec(PMID, "html", link = cosm$URL)) %>%
      kable("html", escape = FALSE) %>%
      kable_styling(bootstrap_options = c("hover", "condensed"))
    cosm$Gene <- as.factor(cosm$Gene)
    lvl <- levels(cosm$Gene)
    for (l in lvl)
    {
      a <- which(cosm$Gene == l)
      if (length(a) != 0)
        t <- t %>% pack_rows(l, min(a), max(a), indent = FALSE)
    }
    print(t)
  } else
    cat("  \n##### No references available.  \n")
}, silent = TRUE)
```


## {-}

```{js, echo=FALSE}
/**
 * Copyright (c) 2007-2015 Ariel Flesler - aflesler<a>gmail<d>com | http://flesler.blogspot.com
 * Licensed under MIT
 * @author Ariel Flesler
 * @version 2.1.2
 */
;(function(f){"use strict";"function"===typeof define&&define.amd?define(["jquery"],f):"undefined"!==typeof module&&module.exports?module.exports=f(require("jquery")):f(jQuery)})(function($){"use strict";function n(a){return!a.nodeName||-1!==$.inArray(a.nodeName.toLowerCase(),["iframe","#document","html","body"])}function h(a){return $.isFunction(a)||$.isPlainObject(a)?a:{top:a,left:a}}var p=$.scrollTo=function(a,d,b){return $(window).scrollTo(a,d,b)};p.defaults={axis:"xy",duration:0,limit:!0};$.fn.scrollTo=function(a,d,b){"object"=== typeof d&&(b=d,d=0);"function"===typeof b&&(b={onAfter:b});"max"===a&&(a=9E9);b=$.extend({},p.defaults,b);d=d||b.duration;var u=b.queue&&1<b.axis.length;u&&(d/=2);b.offset=h(b.offset);b.over=h(b.over);return this.each(function(){function k(a){var k=$.extend({},b,{queue:!0,duration:d,complete:a&&function(){a.call(q,e,b)}});r.animate(f,k)}if(null!==a){var l=n(this),q=l?this.contentWindow||window:this,r=$(q),e=a,f={},t;switch(typeof e){case "number":case "string":if(/^([+-]=?)?\d+(\.\d+)?(px|%)?$/.test(e)){e= h(e);break}e=l?$(e):$(e,q);case "object":if(e.length===0)return;if(e.is||e.style)t=(e=$(e)).offset()}var v=$.isFunction(b.offset)&&b.offset(q,e)||b.offset;$.each(b.axis.split(""),function(a,c){var d="x"===c?"Left":"Top",m=d.toLowerCase(),g="scroll"+d,h=r[g](),n=p.max(q,c);t?(f[g]=t[m]+(l?0:h-r.offset()[m]),b.margin&&(f[g]-=parseInt(e.css("margin"+d),10)||0,f[g]-=parseInt(e.css("border"+d+"Width"),10)||0),f[g]+=v[m]||0,b.over[m]&&(f[g]+=e["x"===c?"width":"height"]()*b.over[m])):(d=e[m],f[g]=d.slice&& "%"===d.slice(-1)?parseFloat(d)/100*n:d);b.limit&&/^\d+$/.test(f[g])&&(f[g]=0>=f[g]?0:Math.min(f[g],n));!a&&1<b.axis.length&&(h===f[g]?f={}:u&&(k(b.onAfterFirst),f={}))});k(b.onAfter)}})};p.max=function(a,d){var b="x"===d?"Width":"Height",h="scroll"+b;if(!n(a))return a[h]-$(a)[b.toLowerCase()]();var b="client"+b,k=a.ownerDocument||a.document,l=k.documentElement,k=k.body;return Math.max(l[h],k[h])-Math.min(l[b],k[b])};$.Tween.propHooks.scrollLeft=$.Tween.propHooks.scrollTop={get:function(a){return $(a.elem)[a.prop]()}, set:function(a){var d=this.get(a);if(a.options.interrupt&&a._last&&a._last!==d)return $(a.elem).stop();var b=Math.round(a.now);d!==b&&($(a.elem)[a.prop](b),a._last=this.get(a))}};return p});
```

```{css, echo=FALSE}
table.dataTable.display tbody tr.highlight {
    background-color: #777777 !important;
    color: #ffffff !important;
}
```

```{js, echo=FALSE}
$(function() {
  $(window).load(function() {
    $('.show-all-start').each(function() {
        var t=$(this);
        t.find('a[role="tab"]').each(function () { $(this).click(); });
        t.find('a[role="tab"]').first().click();
    });
    
    $('a.my_ref_link').click(function () {
        var t = $(this), id=t.data('id'), o=$(id);
        o.parent().parent().addClass('highlight').siblings().removeClass('highlight');
        var tabId = o.closest('.tab-pane').attr('id');
        console.log(tabId);
        o.closest('.tabset').find('a[href="#'+tabId+'"]').click()
        console.log(o.closest('.tabset').find('a[href="#'+tabId+'"]'));
        $('body').scrollTo(id)
    });
  });
});
```
